packages/@lwc/integration-karma/test/directive-external/index.spec.js

```diff
@@ -0,0 +1,76 @@
+import { createElement } from 'lwc';
+import { isNativeShadowRootInstance } from 'test-utils';
+
+import Test from 'x/test';
+
+describe('lwc:external', () => {
+    let test;
+
+    beforeEach(() => {
+        test = createElement('x-test', { is: Test });
+        document.body.appendChild(test);
+    });
+
+    describe('when not upgraded', () => {
+        it('should set only attributes on mount', () => {
+            return Promise.resolve().then(() => {
+                const elm = test.shadowRoot.querySelector('foo-upgrade-never');
+                expect(elm.getAttribute('foo')).toBe('default');
+                expect(elm.foo).toBeUndefined();
+            });
+        });
+
+        it('should set only attributes on update', () => {
+            test.value = 'apple';
+            return Promise.resolve().then(() => {
+                const elm = test.shadowRoot.querySelector('foo-upgrade-never');
+                expect(elm.getAttribute('foo')).toBe('apple');
+                expect(elm.foo).toBeUndefined();
+            });
+        });
+    });
+
+    if (!process.env.COMPAT) {
+        describe('after upgrading', () => {
+            // This test was not broken up into smaller ones with individual assertions because we
+            // cannot manage the order in which different tests run. Order is important because once you
+            // define a custom element, there is no way to undefine it.
+            it('should set property instead of attribute', async () => {
+                const elm = test.shadowRoot.querySelector('foo-upgrade-after');
+
+                test.value = 'sake';
+                await Promise.resolve();
+
+                expect(elm.shadowRoot).toBeNull();
+                expect(elm.getAttribute('foo')).toBe('sake');
+                expect(elm.foo).toBeUndefined();
+
+                test.upgrade('foo-upgrade-after');
+
+                test.value = 'miso';
+                await Promise.resolve();
+
+                expect(elm.getAttribute('foo')).toBe('sake');
+                expect(elm.foo).toBe('miso-prop');
```

Comment:
Just for my understanding: here we have attribute `foo` unchanged because we only set property as the property exists so attribute is not changed. But why the property is changed to `miso-prop` instead of `miso`? This implies that `attributeChangedCallback` is invoked but why it is invoked if we only change the property on the element?

Reply:
The value would be `miso-attr` if it were set as an attribute and the value would be `miso-prop` if it were set as a property.